poleno/mail/tests/test_mandrill.py

```diff
@@ -328,24 +332,20 @@ def _check_response(self, response, klass=HttpResponse, status_code=200, content
         if content is not None:
             self.assertEqual(response.content, content)
 
-    @unittest.skip(u'FIXME')
     def test_allowed_http_methods(self):
         allowed = [u'HEAD', u'GET', u'POST']
         self.assert_allowed_http_methods(allowed, self._webhook_url())
 
-    @unittest.skip(u'FIXME')
     def test_post_method_needs_signature(self):
         with self._overrides():
             response = self.client.post(self._webhook_url(), secure=True)
-        self._check_response(response, HttpResponseForbidden, 403, u'X-Mandrill-Signature not set')
+        self._check_response(response, HttpResponseForbidden, 403)
```

Comment:
@martinmacko47 [docs](https://docs.djangoproject.com/en/1.8/topics/testing/tools/#exceptions), vynimky su neviditelne pre test clienta, a spravanie treba kontrolovat cez `response.status_code`. Atribut content sa zmenil na obsah template `403.html` : `<h1>403 Forbidden</h1>`. Tym padom to nepovazujem za podstatne kontrolovat content. Tiez parameter `content` vo funkcii `_check_response` nie je nikdy pouzity, ale nechal som ho tam.

Reply:
@viliambalaz Chapem. Kazdopadne, tak ci tak treba otestovat, koli comu endpoint padol, nie len ze padol. Inac casom moze zadat endpoint padat z ineho dovodu a my si to nevsimneme, kedze po tento zmene uz neassertujeme konkretny dovod preco endpoint spadol.

Ked zistit dovod, preco endpoint padol, uz teraz nejde z `response.content`, tak to treba otestovat nejak inac. Asi najpraktickejsie bude pozriet sa, co pocas volania endpointu Django zapisalo do loggera.

Ked view raisne `PermissionDenied`, tak Django zapise o tom log do `django.request` loggeru. Ked view raisne `SuspiciousOperation`, tak zapise log do niektoreho z `django.security.*` loggerov. Pricom do zalogovanych logov by sa mal zapisat aj reason string z exception.

Skusme teda spravit pomocny context manager, ktory sa "napichne" na dany logger, a po skonceni contextu vrati zoznam logov, ktore sa pocas beho contextu zalogovali. Analogicky ako mame context manager `created_instances` v `poleno/utils/test.py`, spravme context manager `logged_logs`. 

Skus pouzit postup z tejto podpovede pomocou `mock`: https://stackoverflow.com/questions/14690267/how-do-i-test-if-a-certain-log-message-is-logged-in-a-django-test-case

Reply:
@martinmacko47 Asi je zbytocne vytvarat context manager pre logger. V `test_mandrill.MandrillTransportTest.test_failed_post_request_logs_error`sa pouziva jedno z rieseni:
```
        with mock.patch(u'poleno.mail.cron.cron_logger', logger):
            self._run_mail_cron_job(status_code=404)
        logged = logger.error.call_args[0][0]
```
Avsak rovnake riesenie som nemohol pouzit pre `SuspiciousOperation`, pretoze `security_logger` je definovany vo vnutri funkcii ako lokalna premenna. Zo stackoverflow som preto pouzil riesenie pomocou `patch_logger`. On vlastne splna funkcionalitu ktoru si uviedol vyssie.
Teda `PermissionDenied` testy mam zatial nechat tak? Co mam s nimi teraz spravit?
V novsich verziach je to nahradene spolocnym loggerom `django.request`.

Reply:
@viliambalaz Ano, `patch_logger` je presne to, co som pisal, ze treba spravit. Ani som nevedel, ze django to ma uz hotove, takze jasne, pouzime to. Diki, ze si si to vsimol.

Aby sme v kazdom teste nemuseli zakazdym volat `patch_logger`, presun ho do `self._overrides`. Tam su vsetky overridy, co pri tychto testoch pouzivame. Vysledne `log_messages` z overridu vloz niekam do `self.log_messages` alebo tak nejak.

A na koniec v `self._check_response` nahrad argument `content` argumentom `error`, ktory assertnes, ze je v `self.log_messages`. Alebo zhruba tak nejak.

A co sa tyka `PermissionDenied`, asi by bolo najvhodnejsie upravit implementaciu endpointu, aby namiesto `PermissionDenied` raisoval `SuspiciousOperation` aj v ostatnych situaciach. Kedze aj vsetky tie ostatne situacie (chybajuci resp. zly podpis requesty) su podozrive, ak nastavu v realnej prevadzke.

Reply:
@martinmacko47 `PermissonDenied` som nahradil aj v testoch `test_webhook_secret_with_[custom/default]_name_does_not_match`.